### PR TITLE
Add `array_distance` function

### LINKED ISSUE #8782: Support `array_distance` in `array_expression`
### Is your feature request related to a problem or challenge?

The current implementation of `array_expression.rs` lacks support for the array_distance functionality. This feature is essential for users who want to calculate distances between arrays using specific metrics or custom distance functions.

expect result
```
array_distance([1, 2, 3], [4, 5, 6]) --> 5.196152422706632
array_distance([1, 2, 3], [4, 5, 6], 'l1') --> 3
array_distance([1, 2, 3], [4, 5, 6], 'l2') --> 5.196152422706632
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

list_distance [duckdb](https://duckdb.org/docs/sql/functions/nested)
l1distance [clickhouse](https://clickhouse.com/docs/en/sql-reference/functions/distance-functions#l1distance)l1distance l2distance [clickhouse](https://clickhouse.com/docs/en/sql-reference/functions/distance-functions#l2distance)

### LINKED ISSUE #12210: Add Scalar and Aggregate Distance UDFs
### Is your feature request related to a problem or challenge?

1. Since there are aggregate functions like `corr` and `covar`, it makes sense to me that we should add aggregate `distance`.
2. Beyond the aggregate function, it might also be valuable to add scalar UDFs like `list_distance`/`array_distance`, similar to DuckDB, along with other methods of distance measurement (e.g., like `cosine` etc). Please check https://github.com/apache/datafusion/issues/8782

This function is particularly useful for scenarios involving spatial analysis, clustering, or similarity computations, where distance metrics are crucial. 

### Describe the solution you'd like

XREF: DuckDB [list functions](https://duckdb.org/docs/sql/functions/list.html#list_distancelist1-list2), [array functions](https://duckdb.org/docs/sql/functions/array#array_distancearray1-array2)
> side note: `list` and `array` are different by their own definition.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by austin362667**
## Which issue does this PR close?

Partially closes #8782.

## Rationale for this change

Add **distance functionality** to DataFusion. This function is particularly useful for scenarios involving spatial analysis, clustering, or similarity computations, where distance metrics are crucial. 

It might be valuable to add scalar UDFs like `list_distance`/`array_distance`, similar to DuckDB, along with other methods of distance measurement (e.g., cosine etc)

XREF: DuckDB [list functions](https://duckdb.org/docs/sql/functions/list.html#list_distancelist1-list2), [array functions](https://duckdb.org/docs/sql/functions/array#array_distancearray1-array2)
> side note: In DuckDB, `list` and `array` have different definitions, but in DataFusion, they are considered the same.

## What changes are included in this PR?

New function **Euclidean** `array_distance(arr1, arr2)` is added.

## Are these changes tested?

Yes, added SQL logic tests.

Screenshot 2024-08-29 at 12 20 38 PM

## Are there any user-facing changes?

New function `array_distance(arr1, arr2)` is added.

Examples:
```
> SELECT array_distance([1, 2, 3], [4, 5, 6]);
+-----------------------------------------------------------------------------------------------+
| array_distance(make_array(Int64(1),Int64(2),Int64(3)),make_array(Int64(4),Int64(5),Int64(6))) |
+-----------------------------------------------------------------------------------------------+
| 5.196152422706632                                                                             |
+-----------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.010 seconds.
```
```
> CREATE TABLE points (
    point_a DOUBLE[],
    point_b DOUBLE[]
);
0 row(s) fetched.
Elapsed 0.039 seconds.

> INSERT INTO points VALUES
([1.0, 2.0, 3.0], [1.0, 2.0, 5.0]),
([2.0, 4.0, 6.0], [2.0, 4.0, 6.0]),
([1.5, 2.5, 3.5], [4.5, 6.5, 8.5]);
+-------+
| count |
+-------+
| 3     |
+-------+
1 row(s) fetched.
Elapsed 0.036 seconds.

> SELECT
    point_a,
    point_b,
    list_distance(point_a, point_b) AS euclidean_distance
FROM
    points;
+-----------------+-----------------+--------------------+
| point_a         | point_b         | euclidean_distance |
+-----------------+-----------------+--------------------+
| [1.0, 2.0, 3.0] | [1.0, 2.0, 5.0] | 2.0                |
| [2.0, 4.0, 6.0] | [2.0, 4.0, 6.0] | 0.0                |
| [1.5, 2.5, 3.5] | [4.5, 6.5, 8.5] | 7.0710678118654755 |
+-----------------+-----------------+--------------------+
3 row(s) fetched.
Elapsed 0.013 seconds.

```

No breaking change.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 1]
```diff
@@ -416,6 +416,50 @@ from data
...
+query R
+with data as (
+  select null::int as f, null::int as b
+  union all
+  select null::int as f, null::int as b
+)
+select distance(f, b)
+from data
+----
+0
```
Comment (austin362667) [inline datafusion/sqllogictest/test_files/aggregate.slt:441]: Should here be NULL? If so, I will modify the initial value of `sum_of_squares` in accumulator states too.

File: datafusion/functions-aggregate/src/distance.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+impl Default for Distance {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Distance {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::uniform(2, NUMERICS.to_vec(), Volatility::Immutable),
```
Comment (Weijun-H) [inline datafusion/functions-aggregate/src/distance.rs:63]:
```suggestion
            signature: Signature::numeric(2, Volatility::Immutable),
```

File: datafusion/functions-aggregate/src/distance.rs [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if !arg_types[0].is_numeric() {
+            return plan_err!("Distance requires numeric input types");
+        }
```
Comment (Weijun-H) [inline datafusion/functions-aggregate/src/distance.rs:84]:
```suggestion
        if !arg_types[0].is_numeric() || !arg_types[1].is_numeric() {
            return exec_err!("Distance requires numeric input types");
        }
```

File: datafusion/functions-aggregate/src/distance.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            vec![values1, values2]
+        } else {
+            values.to_vec()
+        };
+
+        let values1 = &cast(&values[0], &DataType::Float64)?;
+        let values2 = &cast(&values[1], &DataType::Float64)?;
+
+        let mut arr1 = downcast_value!(values1, Float64Array).iter().flatten();
+        let mut arr2 = downcast_value!(values2, Float64Array).iter().flatten();
```
Comment (Weijun-H) [inline datafusion/functions-aggregate/src/distance.rs:137]:
```suggestion
        let mut arr1 = as_float64_array(values1)?.iter().flatten();
        let mut arr2 = as_float64_array(values2)?.iter().flatten();
```

File: datafusion/functions-aggregate/src/distance.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    type_coercion::aggregates::NUMERICS,
+    utils::format_state_name,
+    Accumulator, AggregateUDFImpl, Signature, Volatility,
+};
+
+make_udaf_expr_and_func!(
+    Distance,
+    dis,
+    y x,
+    "Distance between two numeric values.",
```
Comment (Weijun-H) [inline datafusion/functions-aggregate/src/distance.rs:45]:
```suggestion
    "Calculates the Euclidean distance between two numeric values.",
```

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 6]
```diff
@@ -416,6 +416,50 @@ from data
...
+  union all
+  select 3 as f,       6 as b
+  union all
+  select null as f, null as b
+)
+select distance(f, b)
+from data
+----
+5.196152422707
+
```
Comment (Weijun-H) [inline datafusion/sqllogictest/test_files/aggregate.slt:462]:
```
> select distance(sq.column1, sq.column2) from (values (NULL, 2), (0,0)) as sq;
+---------------------------------+
| distance(sq.column1,sq.column2) |
+---------------------------------+
| 0.0                             |
+---------------------------------+
```

I prefer it to be `NULL`

**Comment by Weijun-H** [comment]
I am unsure how useful this feature would be because it only calculates two scalar column distances. 🤔

**Comment by jayzhan211** [comment]
@austin362667 I think `array_distance` is not aggregate function but array function. We tends to add new functions if other well-known DBs are also supported. In this case, I think it makes sense to have `array_distance` as array function but not aggregate function

**Comment by austin362667** [comment]
Got it! Thanks for reviewing.
Just adding `list_distance` in functions-nested and removing aggregate `distance` function from functions-aggregate.

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 6 above)
Comment (austin362667) [reply datafusion/sqllogictest/test_files/aggregate.slt]: Sure that's do it this way~

**Review by jayzhan211 — APPROVED** [review]
👍 

We could support column value on the follow up PR

**Comment by Weijun-H** [comment]
Nice! Please add some doc here 
[## Array Functions](https://github.com/austin362667/arrow-datafusion/blob/32d4d8d54e126835cc6065b723b571d967f9c0aa/docs/source/user-guide/sql/scalar_functions.md#L2088-L2089)

File: docs/source/user-guide/sql/scalar_functions.md [hunk 7]
```diff
@@ -2093,6 +2093,7 @@ to_unixtime(expression[, ..., format_n])
 - [array_concat](#array_concat)
 - [array_contains](#array_contains)
 - [array_dims](#array_dims)
+- [array_distance](#array_distance)
```
Comment (Weijun-H) [inline docs/source/user-guide/sql/scalar_functions.md:2096]: also add a bulletin for `list_distance`

Comment (austin362667) [reply docs/source/user-guide/sql/scalar_functions.md:2096]: done! thanks for reminding

**Comment by Weijun-H** [comment]
Thanks, @austin362667, for the contribution, and @jayzhan211 for the review. We'll support the column value in a follow-up PR.

**Pull request merged**